File: go/pkg/basecamp/message_boards_test.go
package basecamp

import (
	"encoding/json"
	"os"
	"path/filepath"
	"testing"
)

func messageBoardsFixturesDir() string {
	return filepath.Join("..", "..", "..", "spec", "fixtures", "message_boards")
}

func loadMessageBoardsFixture(t *testing.T, name string) []byte {
	t.Helper()
	path := filepath.Join(messageBoardsFixturesDir(), name)
	data, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("failed to read fixture %s: %v", name, err)
	}
	return data
}

func TestMessageBoard_UnmarshalGet(t *testing.T) {
	data := loadMessageBoardsFixture(t, "get.json")

	var board MessageBoard
	if err := json.Unmarshal(data, &board); err != nil {
		t.Fatalf("failed to unmarshal get.json: %v", err)
	}

	if board.ID != 1069479338 {
		t.Errorf("expected ID 1069479338, got %d", board.ID)
	}
	if board.Status != "active" {
		t.Errorf("expected status 'active', got %q", board.Status)
	}
	if board.Type != "Message::Board" {
		t.Errorf("expected type 'Message::Board', got %q", board.Type)
	}
	if board.Title != "Message Board" {
		t.Errorf("expected title 'Message Board', got %q", board.Title)
	}
	if board.URL != "https://3.basecampapi.com/195539477/buckets/2085958499/message_boards/1069479338.json" {
		t.Errorf("unexpected URL: %q", board.URL)
	}
	if board.AppURL != "https://3.basecamp.com/195539477/buckets/2085958499/message_boards/1069479338" {
		t.Errorf("unexpected AppURL: %q", board.AppURL)
	}

	// Verify messages count and URL
	if board.MessagesCount != 12 {
		t.Errorf("expected MessagesCount 12, got %d", board.MessagesCount)
	}
	if board.MessagesURL != "https://3.basecampapi.com/195539477/buckets/2085958499/message_boards/1069479338/messages.json" {
		t.Errorf("unexpected MessagesURL: %q", board.MessagesURL)
	}

	// Verify timestamps are parsed
	if board.CreatedAt.IsZero() {
		t.Error("expected CreatedAt to be non-zero")
	}
	if board.UpdatedAt.IsZero() {
		t.Error("expected UpdatedAt to be non-zero")
	}

	// Verify bucket
	if board.Bucket == nil {
		t.Fatal("expected Bucket to be non-nil")
	}
	if board.Bucket.ID != 2085958499 {
		t.Errorf("expected Bucket.ID 2085958499, got %d", board.Bucket.ID)
	}
	if board.Bucket.Name != "The Leto Laptop" {
		t.Errorf("expected Bucket.Name 'The Leto Laptop', got %q", board.Bucket.Name)
	}
	if board.Bucket.Type != "Project" {
		t.Errorf("expected Bucket.Type 'Project', got %q", board.Bucket.Type)
	}

	// Verify creator
	if board.Creator == nil {
		t.Fatal("expected Creator to be non-nil")
	}
	if board.Creator.ID != 1049715914 {
		t.Errorf("expected Creator.ID 1049715914, got %d", board.Creator.ID)
	}
	if board.Creator.Name != "Victor Cooper" {
		t.Errorf("expected Creator.Name 'Victor Cooper', got %q", board.Creator.Name)
	}
	if board.Creator.EmailAddress != "victor@honchodesign.com" {
		t.Errorf("expected Creator.EmailAddress 'victor@honchodesign.com', got %q", board.Creator.EmailAddress)
	}
	if !board.Creator.Owner {
		t.Error("expected Creator.Owner to be true")
	}
	if !board.Creator.Admin {
		t.Error("expected Creator.Admin to be true")
	}
}

File: typescript/tests/services/message-boards.test.ts
/**
 * Tests for the MessageBoardsService class
 */
import { describe, it, expect, beforeEach } from "vitest";
import { http, HttpResponse } from "msw";
import { server } from "../setup.js";
import { createBasecampClient, type BasecampClient } from "../../src/client.js";
import { BasecampError } from "../../src/errors.js";

const BASE_URL = "https://3.basecampapi.com/12345";

describe("MessageBoardsService", () => {
  let client: BasecampClient;

  beforeEach(() => {
    client = createBasecampClient({
      accountId: "12345",
      accessToken: "test-token",
    });
  });

  describe("get", () => {
    it("should get a message board by ID", async () => {
      const mockBoard = {
        id: 123,
        status: "active",
        title: "Message Board",
        created_at: "2024-01-01T00:00:00Z",
        updated_at: "2024-01-02T00:00:00Z",
        type: "Message::Board",
        url: "https://3.basecampapi.com/12345/message_boards/123.json",
        app_url: "https://3.basecamp.com/12345/message_boards/123",
        messages_count: 5,
        messages_url: "https://3.basecampapi.com/12345/message_boards/123/messages.json",
        bucket: { id: 1, name: "Test Project", type: "Project" },
        creator: { id: 999, name: "Test User", email_address: "test@example.com" },
      };

      server.use(
        http.get(`${BASE_URL}/message_boards/123`, () => {
          return HttpResponse.json(mockBoard);
        })
      );

      const board = await client.messageBoards.get(123);

      expect(board.id).toBe(123);
      expect(board.title).toBe("Message Board");
      expect(board.messages_count).toBe(5);
      expect(board.bucket?.id).toBe(1);
      expect(board.creator?.name).toBe("Test User");
    });

    it("should throw not_found for non-existent board", async () => {
      server.use(
        http.get(`${BASE_URL}/message_boards/999`, () => {
          return HttpResponse.json({ error: "Not found" }, { status: 404 });
        })
      );

      try {
        await client.messageBoards.get(999);
        expect.fail("Should have thrown");
      } catch (err) {
        expect(err).toBeInstanceOf(BasecampError);
        expect((err as BasecampError).code).toBe("not_found");
      }
    });
  });
});
